File: python/day6.py
from utils import read_text, file, time_taken


def get_test(part: int):
    return read_text(file(6, part)).strip()


def part1(note: str):
    mentors = 0
    pairs = 0
    for ch in note:
        if ch == "A":
            mentors += 1
        elif ch == "a":
            pairs += mentors
    return pairs


def part2(note: str):
    mentors = [0, 0, 0]
    pairs = 0
    for ch in note:
        if ch.isupper():
            mentors[ord(ch) - ord("A")] += 1
        else:
            pairs += mentors[ord(ch.upper()) - ord("A")]

    return pairs


def part3(note: str):
    rep = 1000
    dist = 1000
    pairs = 0
    for i, ch in enumerate(note):
        if ch.islower():
            mentor = ch.upper()

            for j in range(-dist, dist + 1):
                offset = i + j
                pos = offset if offset < len(note) else offset % len(note)
                if note[pos] == mentor:
                    pairs += rep - (1 if offset < 0 or offset > len(note) else 0)

    return pairs


if __name__ == "__main__":
    print("Part 1:")
    time_taken(lambda: part1(get_test(1)))
    print("Part 2:")
    time_taken(lambda: part2(get_test(2)))
    print("Part 3:")
    time_taken(lambda: part3(get_test(3)))

File: rust/src/day_6.rs
use crate::utils::Solution;
use std::fmt::Display;

pub struct Day6;

impl Solution for Day6 {
    const DAY: u32 = 6;
    fn part1(input: &str) -> impl Display {
        let data = input;
        let mut mentors = 0;
        let mut pairs = 0;
        for ch in data.chars() {
            if ch == 'A' {
                mentors += 1;
            } else if ch == 'a' {
                pairs += mentors;
            }
        }

        pairs
    }

    fn part2(input: &str) -> impl Display {
        let data = input;
        let mut mentors = [0, 0, 0];
        let mut pairs = 0;
        for ch in data.bytes() {
            if ch.is_ascii_uppercase() {
                mentors[(ch - b'A') as usize] += 1;
            } else {
                pairs += mentors[(ch.to_ascii_uppercase() - b'A') as usize];
            }
        }

        pairs
    }

    fn part3(input: &str) -> impl Display {
        let rep = 1000;
        let dist = 1000;
        let data = input.as_bytes();
        let size = data.len();
        let mut pairs = 0;

        for (i, ch) in data.iter().enumerate() {
            if ch.is_ascii_lowercase() {
                let mentors = ch.to_ascii_uppercase();

                for j in -dist..dist + 1 {
                    let offset = i as i64 + j;
                    let wrapped = offset.rem_euclid(size as i64) as usize;

                    if data[wrapped] == mentors {
                        pairs += rep - (offset < 0 || offset > size as i64) as u64;
                    }
                }
            }
        }

        pairs
    }
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn part_1() {
        assert_eq!(Day6::part1("ababacacbcbca").to_string(), "5");
    }

    #[test]
    fn part_2() {
        assert_eq!(Day6::part2("ababacacbcbca").to_string(), "3691");
    }

    #[test]
    fn part_3() {
        assert_eq!(
            Day6::part3("aabcbabcabcabcabcabccbaacbca").to_string(),
            "3442321"
        );
    }
}
